### PR TITLE
Update mod to 1.12

### PR DESCRIPTION
**Pull request by Lothrazar**
I wasnt sure where to send PR into, its pointed at 1.11 branch right now but you will probably want to make your own 1.12 branch.

The testing books work, i opened both up and recipes are updated.  There is a new step of Ingredient objects in the ore dict recipes.

Feel free to leave any feedback, i can make changes if requested.

### CODE REVIEW

File: src/main/java/amerifrance/guideapi/page/reciperenderer/BasicRecipeRenderer.java [hunk 1]
```diff
@@ -64,7 +64,8 @@ public void draw(Book book, CategoryAbstract category, EntryAbstract entry, int
     protected ItemStack getNextItem(ItemStack stack, int position) {
         NonNullList<ItemStack> subItems = NonNullList.create();
         Item item = stack.getItem();
-        item.getSubItems(item, item.getCreativeTab(), subItems);
+         
+        item.getSubItems(  item.getCreativeTab(), subItems);
```
Comment (TehNut) [inline src/main/java/amerifrance/guideapi/page/reciperenderer/BasicRecipeRenderer.java]: Odd spacing

File: src/main/java/amerifrance/guideapi/page/reciperenderer/ShapedOreRecipeRenderer.java [hunk 2]
```diff
@@ -52,8 +57,10 @@ public void draw(Book book, CategoryAbstract category, EntryAbstract entry, int
                         GuiHelper.drawItemStack(input, stackX, stackY);
                         if (GuiHelper.isMouseBetween(mouseX, mouseY, stackX, stackY, 15, 15))
                             tooltips = GuiHelper.getTooltip(input);
-                    } else {
-                        List<ItemStack> list = (List<ItemStack>) component;
+                    }
+                    else if(component instanceof  OreIngredient){
+                      OreIngredient ing = ( OreIngredient) component;
+                      List<ItemStack> list = Arrays.asList(    ing.getMatchingStacks());
```
Comment (TehNut) [inline src/main/java/amerifrance/guideapi/page/reciperenderer/ShapedOreRecipeRenderer.java]: Weird spacing

File: src/main/java/amerifrance/guideapi/page/reciperenderer/ShapedRecipesRenderer.java [hunk 3]
```diff
@@ -27,11 +28,13 @@ public void draw(Book book, CategoryAbstract category, EntryAbstract entry, int
...
                 int stackY = (y + 1) * 17 + (guiTop + 38) + y;
-                ItemStack stack = recipe.recipeItems[y * recipe.recipeWidth + x];
+                Ingredient ing =  recipe.recipeItems.get(y * recipe.recipeWidth + x);
+                if(ing.getMatchingStacks().length==0){continue;}
+                ItemStack stack = ing.getMatchingStacks()[0];
                 if (!stack.isEmpty()) {
                     if (stack.getItemDamage() == OreDictionary.WILDCARD_VALUE) {
                         NonNullList<ItemStack> subItems = NonNullList.create();
-                        stack.getItem().getSubItems(stack.getItem(), stack.getItem().getCreativeTab(), subItems);
+                        stack.getItem().getSubItems( stack.getItem().getCreativeTab(), subItems);
```
Comment (TehNut) [inline src/main/java/amerifrance/guideapi/page/reciperenderer/ShapedRecipesRenderer.java]: Weird spacing

File: src/main/java/amerifrance/guideapi/test/TestBook.java [hunk 4]
```diff
@@ -65,6 +65,6 @@ public void handleModel(ItemStack bookStack) {
 
     @Override
     public void handlePost(ItemStack bookStack) {
-        GameRegistry.addRecipe(new ShapedOreRecipe(bookStack, "X X", " X ", "X X", 'X', "ingotIron"));
+       // GameRegistry.addShapedRecipe(new ResourceLocation("guideapi","testbook1"),bookStack, "X X", " X ", "X X", 'X', "ingotIron");
```
Comment (TehNut) [inline src/main/java/amerifrance/guideapi/test/TestBook.java]: Fix this for new recipe system. I'm not using JSONs

File: src/main/java/amerifrance/guideapi/test/TestBook2.java [hunk 5]
```diff
@@ -64,6 +64,6 @@ public void handleModel(ItemStack bookStack) {
 
     @Override
     public void handlePost(ItemStack bookStack) {
-        GameRegistry.addRecipe(new ShapedOreRecipe(bookStack, "X X", " X ", "X X", 'X', "ingotIron"));
+      //  GameRegistry.addRecipe(new ShapedOreRecipe(bookStack, "X X", " X ", "X X", 'X', "ingotIron"));
```
Comment (TehNut) [inline src/main/java/amerifrance/guideapi/test/TestBook2.java]: Fix this for new recipe system. I'm not using JSONs

**Review by TehNut — CHANGES_REQUESTED** [review]
There's some odd spacing changes throughout, could you do a pass over the diff and look for any unnecessary ones?

**Comment by Lothrazar** [comment]
Yeah sorry about the spacing, i must have been half asleep. also i have some formatting rules from my own projects that are probably trying to apply here.

**Comment by Lothrazar** [comment]
Updated recipes, and I THINK i fixed all the weird spacing issues.

**Comment by TehNut** [comment]
Well, that's definitely one way to format recipes

**Review by TehNut — APPROVED** [review]

**Comment by TehNut** [comment]
Going to cleanup some things then I'll run a build

**Comment by Lothrazar** [comment]
Cool!

**Comment by Lothrazar** [comment]
I found a bug in shapeless recipe . ill make PR later but im just afk for a bit.  When i narrow it down ill also add some more variations of shapeless recipes to one of the Test Books to be safe

java.lang.IndexOutOfBoundsException: Index: 3, Size: 3
	at java.util.ArrayList.rangeCheck(Unknown Source)
	at java.util.ArrayList.get(Unknown Source)
	at net.minecraft.util.NonNullList.get(NonNullList.java:51)
	at amerifrance.guideapi.page.reciperenderer.ShapelessRecipesRenderer.draw(ShapelessRecipesRenderer.java:33)
	at amerifrance.guideapi.page.PageIRecipe.draw(PageIRecipe.java:59)
	at amerifrance.guideapi.wrapper.PageWrapper.draw(PageWrapper.java:50)
	at amerifrance.guideapi.gui.GuiEntry.drawScreen(GuiEntry.java:82)
	at net.minecraftforge.client.ForgeHooksClient.drawScreen(ForgeHooksClient.java:382)
	at net.minecraft.client.renderer.EntityRenderer.updateCameraAndRender(EntityRenderer.java:1168)
	at net.minecraft.client.Minecraft.runGameLoop(Minecraft.java:1192)
	at net.minecraft.client.Minecraft.run(Minecraft.java:436)
	at net.minecraft.client.main.Main.main(Main.java:118)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)

**Pull request merged**